python/ray/data/tests/test_dataset_formats.py

```diff
@@ -791,6 +793,20 @@ def test_read_text(ray_start_regular_shared, tmp_path):
     assert ds.count() == 5
 
 
+def test_read_binary_snappy(ray_start_regular_shared, tmp_path):
+    path = os.path.join(tmp_path, "test_binary_snappy")
+    os.mkdir(path)
+    with open(os.path.join(path, "file"), "wb") as f:
+        byte_str = "hello, world".encode()
+        bytes = BytesIO(byte_str)
+        snappy.stream_compress(bytes, f)
+    ds = ray.data.read_binary_files(
+        path,
+        arrow_open_stream_args=dict(compression="snappy"),
+    )
+    assert sorted(ds.take()) == [byte_str]
```

Comment:
Could you also add a test that doesn't explicitly pass in `compression="snappy"` and instead infers it from the file name? Something like:
```python
def test_read_binary_snappy_inferred(ray_start_regular_shared, tmp_path):
    path = os.path.join(tmp_path, "test_binary_snappy")
    os.mkdir(path)
    with open(os.path.join(path, "file.snappy"), "wb") as f:
        byte_str = "hello, world".encode()
        bytes = BytesIO(byte_str)
        snappy.stream_compress(bytes, f)
    ds = ray.data.read_binary_files(path)
    assert sorted(ds.take()) == [byte_str]
```

Reply:
I was running this test and it failed. I then explored the Codec.detect method in pyarrow and found that it doesn't work for snappy. I think we'll need to make the arrow_open_stream_args a compulsory parameter in case of a snappy file. Alternatively, we can write a small check for the file extension to see if the file is snappy compressed or not.

Reply:
Hi Clark, I tries running the suggested test and it failed. On exploring the Codec.detect api, I found that it is not working for snappy and is returning TypeError. There are 2 ways in which we can solve this issue:
1. Write a small check which checks for the file extension and determines whether it's snappy or not;
2. Make the arrow_open_stream_arg compression compulsory for snappy files

Do suggest how I should proceed on this.

Reply:
>On exploring the Codec.detect method in pyarrow, I found that it is not working for snappy and is returning TypeError.

Ah that's too bad, thanks for checking that!

>Do suggest how I should proceed on this.

I would timebox (1) since this _should_ be pretty straightforward to check for using [pathlib](https://docs.python.org/3/library/pathlib.html):
```python
if compression is None:
    import pyarrow as pa
    import pathlib

    try:
        # If no compression manually given, try to detect compression codec from path.
        compression = pa.Codec.detect(read_path).name
    except (ValueError, TypeError):
        # Arrow's compression inference on the file path doesn't work for Snappy, so
        # we double-check ourselves.
        suffix = pathlib.Path(read_path).suffix
        if suffix and suffix[1:] == "snappy":
            compression = "snappy"
        else:
            compression = None
```